### PR TITLE
Add `@as_model` decorator

### PR DESCRIPTION
**Pull request by theorashid**
Also move `marginal_model` into `pymc_experimental/model`. Hopefully I reorganised the file structure correctly

cc: @ricardoV94 @zaxtax @twiecki

### CODE REVIEW

**Comment by ricardoV94** [comment]
Just a bit worried about name class between the module `model` and the decorator `model`? What about calling the decorator `as_model`?

**Comment by ricardoV94** [comment]
The URL in the docstrings doesn't seem to render correctly: https://pymcio--268.org.readthedocs.build/projects/experimental/en/268/generated/pymc_experimental.model.html

File: pymc_experimental/model/model_api.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    Decorator to provide context to PyMC models declared in a function.
+    This removes all need to think about context managers and lets you separate creating a generative model from using the model.
+
+    Inspired by the [sampled](https://github.com/colcarroll/sampled) decorator for PyMC3.
+
+    Examples
+    --------
+    .. code:: python
+
+        # The following are equivalent
```
Comment (ricardoV94) [inline pymc_experimental/model/model_api.py:17]: I suggest including imports to make the snippet copy-pastable

**Comment by zaxtax** [comment]
> Just a bit worried about name clash between the module `model` and the decorator `model`? What about calling the decorator `as_model`?

Have there been bugs or issues due to this kind of name clash before?

**Comment by ricardoV94** [comment]
> Have there been bugs or issues due to this kind of name clash before?

We had some issues with docs in the past, when a function had the same name as the module

**Comment by theorashid** [comment]
hopefully all good to go now

File: pymc_experimental/model/model_api.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import pymc as pm
+
+
+def as_model(*model_args, **model_kwargs):
+    R"""
+    Decorator to provide context to PyMC models declared in a function.
+    This removes all need to think about context managers and lets you separate creating a generative model from using the model.
+
+    Inspired by the `sampled <https://github.com/colcarroll/sampled>`_ decorator for PyMC3.
```
Comment (twiecki) [inline pymc_experimental/model/model_api.py:11]: @zaxtax's blog post was the more direct inspiration here, no? Maybe instead you could link to the blog post.

**Comment by twiecki** [comment]
Would be nice to give @zaxtax some credit here, commit co-author?

File: pymc_experimental/model/model_api.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+from functools import wraps
+
+import pymc as pm
```
Comment (ricardoV94) [inline pymc_experimental/model/model_api.py:3]: Nitpick

```suggestion
from pymc import Model
```

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Thanks @theorashid

File: pymc_experimental/model/model_api.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            pm.Normal("x", 0., 1., dims="obs")
+
+        m = basic_model()
+        pm.sample(model=m)
+
+    """
+
+    def decorator(f):
+        @wraps(f)
+        def make_model(*args, **kwargs):
```
Comment (ferrine) [inline pymc_experimental/model/model_api.py:39]: I would also add a name kwarg here to change model name

Comment (ricardoV94) [reply pymc_experimental/model/model_api.py:39]: You can pass it to model_kwargs

**Pull request merged**